java-checks/src/main/java/org/sonar/java/checks/StringLiteralDuplicatedCheck.java

```diff
@@ -33,57 +32,82 @@
+  // String literals include quotes, so this means length 5 as defined in RSPEC
+  private static final int MINIMAL_LITERAL_LENGTH = 7;
+
   @RuleProperty(
     key = "threshold",
     description = "Number of times a literal must be duplicated to trigger an issue",
     defaultValue = "" + DEFAULT_THRESHOLD)
   public int threshold = DEFAULT_THRESHOLD;
 
   private final Multimap<String, LiteralTree> occurrences = ArrayListMultimap.create();
+  private final Map<String, VariableTree> constants = new HashMap<>();
 
   @Override
   public void scanFile(JavaFileScannerContext context) {
     occurrences.clear();
+    constants.clear();
     scan(context.getTree());
     for (String entry : occurrences.keySet()) {
       Collection<LiteralTree> literalTrees = occurrences.get(entry);
-      int literalOccurence = literalTrees.size();
-      if (literalOccurence >= threshold) {
-        List<JavaFileScannerContext.Location> flow = new ArrayList<>();
-        for (Tree element : literalTrees) {
-          flow.add(new JavaFileScannerContext.Location("Duplication", element));
-        }
+      int literalOccurrence = literalTrees.size();
+      if (constants.containsKey(entry) && literalOccurrence > 1) {
+        VariableTree constant = constants.get(entry);
+        List<LiteralTree> duplications = literalTrees.stream().filter(literal -> literal.parent() != constant).collect(Collectors.toList());
+        context.reportIssue(this, duplications.iterator().next(),
+          "Use already defined constant '" + constant.simpleName() + "' instead of duplicating literal " + entry + ".",
+          secondaryLocations(duplications.subList(1, duplications.size())), literalOccurrence);
+      } else if (literalOccurrence >= threshold) {
         context.reportIssue(
           this,
-          Iterables.getFirst(literalTrees, null),
-          "Define a constant instead of duplicating this literal " + entry + " " + literalOccurence + " times.",
-          flow,
-          literalOccurence);
+          literalTrees.iterator().next(),
+          "Define a constant instead of duplicating this literal " + entry + " " + literalOccurrence + " times.",
+          secondaryLocations(literalTrees), literalOccurrence);
       }
     }
   }
 
+  private static List<JavaFileScannerContext.Location> secondaryLocations(Collection<LiteralTree> literalTrees) {
+    return literalTrees.stream().map(element -> new JavaFileScannerContext.Location("Duplication", element)).collect(Collectors.toList());
+  }
+
   @Override
   public void visitLiteral(LiteralTree tree) {
     if (tree.is(Tree.Kind.STRING_LITERAL)) {
       String literal = tree.value();
+      isConstant(tree).ifPresent(constant -> constants.putIfAbsent(literal, constant));
       if (literal.length() >= MINIMAL_LITERAL_LENGTH) {
         occurrences.put(literal, tree);
       }
     }
   }
 
+  private static Optional<VariableTree> isConstant(LiteralTree tree) {
```

Comment:
I think it would have been easier to simply visit variableTrees directly, and once checked that they are `static` and `final`, just check for an STRING_LITERAL initializer.